pyrelational/models/ensemble_model.py

```diff
@@ -1,41 +1,37 @@
 from abc import ABC
-from typing import Any, Dict, List, Optional, Type, Union
+from typing import Any, Dict, Generic, List, Optional, Type, TypeVar, Union
 
 import numpy as np
 import torch
 from pytorch_lightning import LightningModule
 from pytorch_lightning.utilities.model_helpers import is_overridden
+from torch.nn import Module
 from torch.utils.data import DataLoader
 
 from .abstract_model_manager import ModelManager
 from .lightning_model import LightningModel
 from .model_utils import _determine_device
 
+ModelType = TypeVar("ModelType", bound=Module)
 
-class EnsembleManager(ModelManager, ABC):
+
+class EnsembleManager(Generic[ModelType], ModelManager[ModelType, List[ModelType]], ABC):
     """
     Generic wrapper for ensemble uncertainty estimator
     """
 
     def __init__(
         self,
-        model_class: Type[Any],
-        model_config: Union[str, Dict],
-        trainer_config: Union[str, Dict],
+        model_class: Type[ModelType],
+        model_config: Union[str, Dict[str, Any]],
+        trainer_config: Union[str, Dict[str, Any]],
         n_estimators: int = 10,
     ):
         super(EnsembleManager, self).__init__(model_class, model_config, trainer_config)
         self.device = _determine_device(self.trainer_config.get("gpus", 0))
         self.n_estimators = n_estimators
 
-    def init_model(self) -> List[Any]:
```

Comment:
Why remove this btw? I think it makes sense to have a function which initialises the model in this way?

Reply:
hmm I suppose it breaks liskov? but then so do other aspects, like self.current_model being a list, it's just more hidden so Mypy doesn't complain

Reply:
self.current doesn't break it anymore, because of the use of Generic :)

Reply:
which probably means I could add this function back, but I am not sure what we gain from it

Reply:
it is actually a problem, because the parent init_model returns a ModelType and not an E type, while the child class would need to. I don't have a clean way around this :(